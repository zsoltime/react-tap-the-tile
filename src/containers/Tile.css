.tile {
  align-items: stretch;
  background-color: #fff;
  border: 1px solid rgba(0, 0, 0, 0.8);
  cursor: pointer;
  display: flex;
  flex-direction: column;
  height: 5rem;
  justify-content: space-between;
  transition: background 0.3s ease-in-out, color 0.2s ease-in-out;
  width: 5rem;
}

.tile--isActive {
  background-color: #333;
  color: rgba(255, 255, 255, 0.95);
}

.tile__clicks {
  font-size: 2.5rem;
}
.tile__time {
  font-size: 0.825rem;
}
