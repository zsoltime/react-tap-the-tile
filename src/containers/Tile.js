import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledTile = styled.button.attrs({ type: 'button' })`
  align-items: stretch;
  background-color: ${props => (props.isActive ? '#363636' : '#fff')};
  border: 1px solid rgba(0, 0, 0, 0.8);
  color: rgba(255, 255, 255, 0.95);
  cursor: pointer;
  display: flex;
  flex-direction: column;
  height: 23%;
  justify-content: space-between;
  transition: background 0.3s ease-in-out;
  width: 23%;
`;

const RemainingClicks = styled.div`
  font-size: 2.5rem;
`;

const RemainingTime = styled.div`
  font-size: 0.825rem;
`;

const DEFAULT_TIMER = 5000;
const TIMER_INTERVAL = 500;

export class Tile extends Component {
  static propTypes = {
    clicksRemaining: PropTypes.number,
    id: PropTypes.string.isRequired,
    isActive: PropTypes.bool,
    onClick: PropTypes.func.isRequired,
    onTimeout: PropTypes.func.isRequired,
  };

  static defaultProps = {
    clicksRemaining: 0,
    isActive: false,
  };

  state = {
    timeRemaining: DEFAULT_TIMER,
  };

  componentDidUpdate(prevProps) {
    // if it wasn't active but it is now
    if (!prevProps.isActive && this.props.isActive) {
      this.startTimer();
    }
    // if it was active but no longer is
    if (prevProps.isActive && !this.props.isActive) {
      this.stopTimer();
    }
  }

  componentWillUnmount() {
    clearInterval(this.intervalID);
  }

  startTimer = () => {
    if (!this.props.isActive) {
      return;
    }

    this.intervalID = window.setInterval(() => {
      if (
        !this.props.isActive ||
        this.state.timeRemaining <= TIMER_INTERVAL
      ) {
        this.stopTimer();
        this.props.onTimeout(this.props.id);
        return;
      }

      this.setState(state => ({
        timeRemaining: state.timeRemaining - TIMER_INTERVAL,
      }));
    }, TIMER_INTERVAL);
  };

  stopTimer = () => {
    clearInterval(this.intervalID);

    this.setState({
      timeRemaining: DEFAULT_TIMER,
    });
  };

  render() {
    const { timeRemaining } = this.state;
    const { clicksRemaining, id, isActive, onClick } = this.props;

    return (
      <StyledTile isActive={isActive} onClick={() => onClick(id)}>
        {isActive && (
          <>
            <RemainingClicks key="0">
          {clicksRemaining > 1 && clicksRemaining}
            </RemainingClicks>
            <RemainingTime key="1">
          {timeRemaining > 0 && Math.floor(timeRemaining / 1000)}
            </RemainingTime>
          </>
        )}
      </StyledTile>
    );
  }
}

export default Tile;
