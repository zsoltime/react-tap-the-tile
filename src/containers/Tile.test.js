import React from 'react';
import { render, shallow } from 'enzyme';
import 'jest-styled-components';

import Tile, { StyledTile } from './Tile';

describe('<Tile />', () => {
  it('renders button with remaining time and clicks left', () => {
    const wrapper = render(
      <Tile
        clicksRemaining={5}
        id="id"
        isActive
        onClick={f => f}
        onTimeout={f => f}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders button with remaining time if only one click is left', () => {
    const wrapper = render(
      <Tile
        clicksRemaining={1}
        id="id"
        isActive
        onClick={f => f}
        onTimeout={f => f}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders button with the correct remaining time after 2.5 seconds', () => {
    jest.useFakeTimers();
    const timePassed = 2500;
    const expectedTimeRemaining = 3000;
    const expectedSeconds = '3';
    const wrapper = shallow(
      <Tile
        clicksRemaining={1}
        id="id"
        isActive
        onClick={f => f}
        onTimeout={f => f}
      />
    );

    expect(wrapper.state('timeRemaining')).toBe(5000);

    wrapper.instance().startTimer();
    jest.advanceTimersByTime(timePassed);
    expect(wrapper.state('timeRemaining')).toBe(
      expectedTimeRemaining
    );

    expect(
      wrapper
        .render()
        .find('div')
        .last()
        .text()
    ).toBe(expectedSeconds);
  });

  it('renders empty button if inactive', () => {
    const wrapper = render(
      <Tile id="id" onClick={f => f} onTimeout={f => f} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('calls onClick on button click', () => {
    const handleClick = jest.fn();
    const tileId = 'test id';
    const timeRemaining = 5000;

    const wrapper = shallow(
      <Tile id={tileId} onClick={handleClick} onTimeout={f => f} />
    );

    wrapper.find(StyledTile).simulate('click');

    expect(handleClick).toBeCalledTimes(1);
    expect(handleClick).toBeCalledWith(tileId, timeRemaining);
  });

  it('starts a timer if tile becomes active', () => {
    const wrapper = shallow(
      <Tile
        clicksRemaining={1}
        id="id"
        onClick={f => f}
        onTimeout={f => f}
      />
    );

    const startTimer = jest.spyOn(wrapper.instance(), 'startTimer');

    expect(startTimer).not.toBeCalled();

    wrapper.setProps({ isActive: true });
    expect(startTimer).toBeCalledTimes(1);
  });

  it('stops timer if tile becomes inactive', () => {
    const wrapper = shallow(
      <Tile
        clicksRemaining={1}
        id="id"
        isActive
        onClick={f => f}
        onTimeout={f => f}
      />
    );

    const stopTimer = jest.spyOn(wrapper.instance(), 'stopTimer');

    expect(stopTimer).not.toBeCalled();

    wrapper.setProps({ isActive: false });
    expect(stopTimer).toBeCalledTimes(1);
  });

  it('calls onTimeout after five seconds', () => {
    jest.useFakeTimers();
    const handleTimeout = jest.fn();
    const tileId = 'test id';
    const wrapper = shallow(
      <Tile
        clicksRemaining={1}
        id={tileId}
        isActive
        onClick={f => f}
        onTimeout={handleTimeout}
      />
    );

    expect(handleTimeout).not.toBeCalled();
    expect(wrapper.state('timeRemaining')).toBe(5000);

    wrapper.instance().startTimer();
    jest.advanceTimersByTime(2500);
    expect(wrapper.state('timeRemaining')).toBe(3000);

    jest.advanceTimersByTime(2500);
    expect(handleTimeout).toBeCalledTimes(1);
    expect(handleTimeout).toBeCalledWith(tileId);
  });
});
