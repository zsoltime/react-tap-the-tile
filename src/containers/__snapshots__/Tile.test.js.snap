// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> renders button with remaining time and clicks left 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background-color: #363636;
  border: 1px solid rgba(0,0,0,0.8);
  color: rgba(255,255,255,0.95);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 23%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  width: 23%;
}

.c1 {
  font-size: 2.5rem;
}

.c2 {
  font-size: 0.825rem;
}

<button
  class="c0"
  type="button"
>
  <div
    class="c1"
  >
    5
  </div>
  <div
    class="c2"
  >
    5
  </div>
</button>
`;

exports[`<Tile /> renders button with remaining time if only one click is left 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background-color: #363636;
  border: 1px solid rgba(0,0,0,0.8);
  color: rgba(255,255,255,0.95);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 23%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  width: 23%;
}

.c1 {
  font-size: 2.5rem;
}

.c2 {
  font-size: 0.825rem;
}

<button
  class="c0"
  type="button"
>
  <div
    class="c1"
  />
  <div
    class="c2"
  >
    5
  </div>
</button>
`;

exports[`<Tile /> renders empty button if inactive 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  background-color: #fff;
  border: 1px solid rgba(0,0,0,0.8);
  color: rgba(255,255,255,0.95);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 23%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  width: 23%;
}

<button
  class="c0"
  type="button"
/>
`;
